c2rust-analyze/src/rewrite/expr/hir_op.rs

```diff
@@ -127,60 +155,51 @@ impl<'a, 'tcx> HirRewriteVisitor<'a, 'tcx> {
             }
             hir::ExprKind::Path(hir::QPath::Resolved(_, p)) if matches!(p.res, Res::Local(..)) => {
                 // Currently we only handle cases where the local is copied into a temporary.  If
                 // the local is used as-is in some other expression, this case will return `None`.
-                let opt_assign_loc = self
-                    .find_optional_location_matching(
-                        ex.span,
-                        |stmt| matches!(stmt.kind, mir::StatementKind::Assign(..)),
-                        |_term| false,
-                    )
-                    .unwrap_or_else(|err| panic_sole_location_error(err, "Assign statement"));
-                opt_assign_loc
+                push_assign_loc(ex.span)
             }
-            hir::ExprKind::Field(..) => {
-                // Currently we only handle the case where the value retrieved from the field is
-                // stored into a temporary.  If it's stored into a local or some other place (e.g.
-                // `let y = x.f;`, or `y = x.f;` alone), then this case will return `None`.
+            hir::ExprKind::Unary(UnOp::Deref, ..)
+            | hir::ExprKind::Lit(..)
+            | hir::ExprKind::Field(..) => {
+                // For `hir::ExprKind::Field` we currently we only handle the case where the value
+                // retrieved from the field is stored into a temporary.  If it's stored into a
+                // local or some other place (e.g. `let y = x.f;`, or `y = x.f;` alone), then
+                // this case will return `None`.
                 //
                 // Also, for chained field accesses, this only matches the outermost ones.  Code
                 // like `x.0.0` translates into a single MIR statement with the span of the overall
                 // expression.
-                let opt_assign_loc = self
-                    .find_optional_location_matching(
+                push_assign_loc(ex.span)
+            }
+            hir::ExprKind::AddrOf(..) => {
+                locations.extend(
+                    self.find_all_locations_matching(
                         ex.span,
                         |stmt| matches!(stmt.kind, mir::StatementKind::Assign(..)),
                         |_term| false,
                     )
-                    .unwrap_or_else(|err| panic_sole_location_error(err, "Assign statement"));
-                opt_assign_loc
+                    .iter(),
+                );
             }
-            hir::ExprKind::AddrOf(..) => {
-                if ex.span == ex.span.source_callsite() {
-                    // FIXME: this is a hacky proxy for checking if &raw comes from
-                    // std::ptr::addr_of(_mut)! or is compiler-generated. If the
-                    // spans are equal, it was most likely compiler-generated and so we
-                    // skip this particular type of expression for rewriting
-                    return None;
-                }
```

Comment:
How does `find_all_locations_matching` with a `mir::StatementKind::Assign` pattern predicate make the distinction between compiler-generated `&raw` and ones corresponding to a source-level `addr_of[_mut]!` call? This seems subtle enough to merit a comment.

Reply:
It currently does not make a distinction. Before a distinction was necessary because `find_sole_location` would be trying to get only one location where an HIR statement could generate multiple (this line skipped the second and assumed it was a compiler-generated thing based on the span, but it wasn't very robust and wasn't in an ideal place in terms of logic).